src/sentry/runner/commands/backup.py

```diff
@@ -16,9 +16,21 @@
 def import_(src):
     "Imports data from a Sentry export."
 
-    for obj in serializers.deserialize("json", src, stream=True, use_natural_keys=True):
-        if obj.object._meta.app_label not in EXCLUDED_APPS:
-            obj.save()
+    try:
+        with transaction.atomic():
+            for obj in serializers.deserialize("json", src, stream=True, use_natural_keys=True):
+                if obj.object._meta.app_label not in EXCLUDED_APPS:
+                    obj.save()
+    # There are two integrity error exceptions we handle here.
+    # 1. Foreign key check failure
+    # 2. Duplicate key violation
+    # For both, let's display a warning before reraising error.
```

Comment:
I still feel like this is not a great comment. A foreign key check failure means that we are not including a table in the export that we should be. It's a bug if a user hits that. The duplicate key violation is the one that indicates user error. Also, are there other possible integrity errors? I don't know.

Honestly the best user experience would be to have separate error messages for the two cases.

- if (foreign key fail)
    - "File a bug report, it's probably our fault."
- else
    - "Are you restoring from a clean database?" etc.

Reply:
hm I'm wrong here, these are all the failures:
```
INTEGRITY_CONSTRAINT_VIOLATION = '23000'
RESTRICT_VIOLATION = '23001'
NOT_NULL_VIOLATION = '23502'
FOREIGN_KEY_VIOLATION = '23503'
UNIQUE_VIOLATION = '23505'
CHECK_VIOLATION = '23514'
EXCLUSION_VIOLATION = '23P01'
```

Reply:
I'm not sure if foreign key check will always mean it's not on the user. I think this could also potentially happen between backup/restore of different versions due to database changes as seen [here](https://github.com/getsentry/self-hosted/issues/1598). I think the best case is to display a warning of recommended steps such as

 `Are you restoring from a clean database? Using the same version to backup/restore? If yes to both, file a bug report`.

Reply:
> these are all the failures

Nice find. Is that from `psycopg2` docs? Code? ... ?

> this could also potentially happen between backup/restore of different versions

Hrm, okay, seems plausible. In that case how about we put some effort into making sure that all the right tables are included in the export, and that we have some process or helper in place to ensure that engineers add _future_ metadata tables to the export, and then we can keep the error message simple here (don't need to be explicit about filing a bug report since we won't be expecting it really and ofc users will file bug reports if they do hit an actual bug).

Reply:
yep, psycopg2
https://github.com/psycopg/psycopg2/blob/2_8_6/lib/errorcodes.py#L234-L240

Reply:
Are engineers required to specificy `__include_in_export__` explicitly? That would minimize the risk that we forget a table in the future, leading to a foreign key failure. My hunch is that `__include_in_export__` defaults to `false`.

Reply:
yeah, they are required as seen here
https://github.com/getsentry/sentry/blob/master/src/sentry/db/models/base.py#L133

Reply:
The goal is that our tests can more properly find out if the backup/restore workflow works if new tables are added in the future, to minimize that risk of forgetting tables

Reply:
Awesome, glad to be wrong! 😅